### ISSUE TITLE
OpenJ9 0.9.0 + Metronome GC started, but never stopped

### ISSUE THREAD

**Issue by akrus** [eclipse/openj9#3750]
Hi,

I have some weird behavior of  GC with Metronome. Here is tail of the log:

```
<trigger-start id="10870" timestamp="2018-11-20T17:55:54.202" />
<cycle-start id="10871" type="global" contextid="0" timestamp="2018-11-20T17:55:54.203" intervalms="164189.183" />
<gc-op id="10872" type="heartbeat" contextid="10871" timestamp="2018-11-20T17:55:55.210">
  <quanta quantumCount="138" quantumType="mark" minTimeMs="0.190" meanTimeMs="1.135" maxTimeMs="10.127" maxTimestampMs="298234506.467" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.046" maxTimeMs="0.297" />
  <free-mem type="heap" minBytes="90353607696" meanBytes="90929847829" maxBytes="91266826600" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10873" type="heartbeat" contextid="10871" timestamp="2018-11-20T17:55:56.219">
  <quanta quantumCount="203" quantumType="mark" minTimeMs="0.837" meanTimeMs="1.078" maxTimeMs="1.187" maxTimestampMs="298235731.389" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.020" maxTimeMs="0.170" />
  <free-mem type="heap" minBytes="89883596208" meanBytes="90115669810" maxBytes="90350243488" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10874" type="heartbeat" contextid="10871" timestamp="2018-11-20T17:55:57.227">
  <quanta quantumCount="205" quantumType="mark" minTimeMs="0.930" meanTimeMs="1.082" maxTimeMs="1.222" maxTimestampMs="298237478.562" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.014" maxTimeMs="0.150" />
  <free-mem type="heap" minBytes="89486559936" meanBytes="89706503300" maxBytes="89881146448" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10875" type="heartbeat" contextid="10871" timestamp="2018-11-20T17:55:58.103">
  <quanta quantumCount="173" quantumType="mark" minTimeMs="0.700" meanTimeMs="1.084" maxTimeMs="1.261" maxTimestampMs="298238400.795" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.019" maxTimeMs="0.548" />
  <references type="soft" cleared="0" dynamicThreshold="29" maxThreshold="32" />
  <references type="phantom" cleared="80" />
  <finalization enqueued="198" />
  <free-mem type="heap" minBytes="89086454280" meanBytes="89260575243" maxBytes="89486159232" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10876" type="heartbeat" contextid="10871" timestamp="2018-11-20T17:55:59.110">
  <quanta quantumCount="213" quantumType="sweep" minTimeMs="0.609" meanTimeMs="1.019" maxTimeMs="4.578" maxTimestampMs="298238407.754" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.022" maxTimeMs="0.878" />
  <references type="soft" cleared="0" dynamicThreshold="29" maxThreshold="32" />
  <references type="phantom" cleared="17040" />
  <finalization enqueued="42174" />
  <free-mem type="heap" minBytes="89085624672" meanBytes="102117735345" maxBytes="115475093048" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
```

For some reason there is no cycle end. Service stopped working exactly at that time and wasn't responding to any signals, so had to kill it after 30 minutes of doing nothing.

This happened just once.

**Comment by DanHeidinga** [comment]
FYI @amicic

**Comment by pshipton** [comment]
@akrus note that OpenJ9 0.11.0 is released now, although I don't know if it would help with this.
Without any diagnostics, such as a core file, we may not be able to do much. If it happens again, you may be able to use gcore to capture a core file.

**Comment by dmitripivkine** [comment]
If it happen again it might be useful to collect native stacks for all threads for "hanged" process a few times - if it is possible. On Linux it might be done by connecting/disconnecting GDB to/from process directly.
Please note that:
- it is possible that process actually was not hanged but active with some unexpectedly long operations. In this case snapshots for thread stacks will be different
- It is possible that operation of disconnect of debugger will wake up hanged process. We observed this behaviour before for cases of missed OS write barriers (for example see famous "futex" bug for Linux kernel)
- I understand using of GDB might be inconvenient but it is easier than collect and transfer system core size of hundreds of GB.
If you need help with GDB please let me know.

**Comment by akrus** [comment]
Just happened again on another server. Unfortunately wasn't able to get a dump at this time again or check the trick with gdb as my colleague came earlier and killed the process.

GC log stops after:
```
<trigger-start id="10287" timestamp="2018-12-05T20:18:00.889" />
<cycle-start id="10288" type="global" contextid="0" timestamp="2018-12-05T20:18:00.889" intervalms="1087488.999" />
<gc-op id="10289" type="heartbeat" contextid="10288" timestamp="2018-12-05T20:18:01.892">
  <quanta quantumCount="172" quantumType="mark" minTimeMs="0.476" meanTimeMs="1.573" maxTimeMs="19.894" maxTimestampMs="393154372.976" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.003" maxTimeMs="0.076" />
  <free-mem type="heap" minBytes="257587392488" meanBytes="257633871292" maxBytes="257697174848" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10290" type="heartbeat" contextid="10288" timestamp="2018-12-05T20:18:02.146">
  <quanta quantumCount="48" quantumType="mark" minTimeMs="0.816" meanTimeMs="1.385" maxTimeMs="1.640" maxTimestampMs="393155554.759" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.001" maxTimeMs="0.003" />
  <finalization enqueued="250" />
  <free-mem type="heap" minBytes="257576258072" meanBytes="257580510716" maxBytes="257586982984" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10291" type="heartbeat" contextid="10288" timestamp="2018-12-05T20:18:03.152">
  <quanta quantumCount="186" quantumType="sweep" minTimeMs="1.049" meanTimeMs="1.444" maxTimeMs="5.163" maxTimestampMs="393155630.502" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.002" maxTimeMs="0.063" />
  <finalization enqueued="46500" />
  <free-mem type="heap" minBytes="257576258072" meanBytes="263889985296" maxBytes="273312047672" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10292" type="heartbeat" contextid="10288" timestamp="2018-12-05T20:18:04.166">
  <quanta quantumCount="190" quantumType="sweep" minTimeMs="1.066" meanTimeMs="1.416" maxTimeMs="1.632" maxTimestampMs="393156900.653" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.002" maxTimeMs="0.067" />
  <finalization enqueued="47500" />
  <free-mem type="heap" minBytes="273404885040" meanBytes="283753905591" maxBytes="294106366320" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
<gc-op id="10293" type="heartbeat" contextid="10288" timestamp="2018-12-05T20:18:05.174">
  <quanta quantumCount="186" quantumType="sweep" minTimeMs="1.153" meanTimeMs="1.434" maxTimeMs="1.609" maxTimestampMs="393158559.921" />
  <exclusiveaccess-info minTimeMs="0.001" meanTimeMs="0.003" maxTimeMs="0.069" />
  <finalization enqueued="46500" />
  <free-mem type="heap" minBytes="294258485232" meanBytes="304354559244" maxBytes="314539434160" />
  <thread-priority maxPriority="11" minPriority="11" />
</gc-op>
```

**Comment by akrus** [comment]
Happened again, got dump this time. gdb attach/detach didn't work, here are some details:

```
(gdb) where                                                                                                                                                                                                                                  
#0  0x00007f8020e8fd2d in __GI___pthread_timedjoin_ex (threadid=140188270229248, thread_return=thread_return@entry=0x7ffc67465a48, abstime=abstime@entry=0x0, block=block@entry=true) at pthread_join_common.c:89                            
#1  0x00007f8020e8fb5c in __pthread_join (threadid=<optimized out>, thread_return=thread_return@entry=0x7ffc67465a48) at pthread_join.c:24                                                                                                   
#2  0x00007f8020c7da35 in ContinueInNewThread0 (continuation=continuation@entry=0x7f8020c78870 <JavaMain>, stack_size=0, args=args@entry=0x7ffc67465ae0)                                                                                     
    at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/solaris/bin/java_md_solinux.c:1044                                                                                                               
#3  0x00007f8020c7a347 in ContinueInNewThread (ifn=ifn@entry=0x7ffc67465be0, threadStackSize=<optimized out>, argc=<optimized out>, argv=0xbde648, mode=mode@entry=0, what=what@entry=0x0, ret=0)                                            
    at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/share/bin/java.c:2033                                                                                                                            
#4  0x00007f8020c7daeb in JVMInit (ifn=ifn@entry=0x7ffc67465be0, threadStackSize=<optimized out>, argc=<optimized out>, argv=<optimized out>, mode=0, mode@entry=1, what=0x0, what@entry=0x7ffc6746bf37 "myapp1",             
    ret=ret@entry=0) at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/solaris/bin/java_md_solinux.c:1091                                                                                              
#5  0x00007f8020c7b028 in JLI_Launch (argc=0, argv=0xbde648, jargc=<optimized out>, jargv=<optimized out>, appclassc=1, appclassv=0x0, fullversion=0x4007e9 "1.8.0_181-b13", dotversion=0x4007e5 "1.8", pname=0x4007e0 "java",               
    lname=0x4007d8 "openjdk", javaargs=0 '\000', cpwildcard=1 '\001', javaw=0 '\000', ergo=0) at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/share/bin/java.c:304                                   
#6  0x000000000040060c in main ()
```

```
(gdb) bt full
#0  0x00007f8020e8fd2d in __GI___pthread_timedjoin_ex (threadid=140188270229248, thread_return=thread_return@entry=0x7ffc67465a48, abstime=abstime@entry=0x0, block=block@entry=true) at pthread_join_common.c:89                            
        __tid = 31274                                                                                                                                                                                                                        
        _buffer = {__routine = 0x7f8020e8fb90 <cleanup>, __arg = 0x7f80200c7d28, __canceltype = 0, __prev = 0x0}                                                                                                                             
        oldtype = 0                                                                                                                                                                                                                          
        pd = 0x7f80200c7700                                                                                                                                                                                                                  
        self = <optimized out>                                                                                                                                                                                                               
        result = 0                                                                                                                                                                                                                           
#1  0x00007f8020e8fb5c in __pthread_join (threadid=<optimized out>, thread_return=thread_return@entry=0x7ffc67465a48) at pthread_join.c:24                                                                                                   
No locals.
#2  0x00007f8020c7da35 in ContinueInNewThread0 (continuation=continuation@entry=0x7f8020c78870 <JavaMain>, stack_size=0, args=args@entry=0x7ffc67465ae0)
    at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/solaris/bin/java_md_solinux.c:1044
        tmp = 0xbdffe0                                                                                                                                                                                                                       
        rslt = <optimized out>                                                                                                                                                                                                               
        tid = 140188270229248                                                                                                                                                                                                                
        attr = {__size = '\000' <repeats 17 times>, "\020", '\000' <repeats 37 times>, __align = 0}                                                                                                                                          
#3  0x00007f8020c7a347 in ContinueInNewThread (ifn=ifn@entry=0x7ffc67465be0, threadStackSize=<optimized out>, argc=<optimized out>, argv=0xbde648, mode=mode@entry=0, what=what@entry=0x0, ret=0)                                            
    at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/share/bin/java.c:2033                                                                                                                            
        args = {argc = 0, argv = 0xbde648, mode = 1, what = 0x7ffc6746bf37 "myapp1", ifn = {CreateJavaVM = 0x7f80204708e0 <JNI_CreateJavaVM>, GetDefaultJavaVMInitArgs = 0x7f8020471880 <JNI_GetDefaultJavaVMInitArgs>,       
            GetCreatedJavaVMs = 0x7f8020471860 <JNI_GetCreatedJavaVMs>}}                                                                                                                                                                     
        rslt = <optimized out>
#4  0x00007f8020c7daeb in JVMInit (ifn=ifn@entry=0x7ffc67465be0, threadStackSize=<optimized out>, argc=<optimized out>, argv=<optimized out>, mode=0, mode@entry=1, what=0x0, what@entry=0x7ffc6746bf37 "myapp1",             
    ret=ret@entry=0) at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/solaris/bin/java_md_solinux.c:1091                                                                                              
No locals.                                                                                                                                                                                                                                   
#5  0x00007f8020c7b028 in JLI_Launch (argc=0, argv=0xbde648, jargc=<optimized out>, jargv=<optimized out>, appclassc=1, appclassv=0x0, fullversion=0x4007e9 "1.8.0_181-b13", dotversion=0x4007e5 "1.8", pname=0x4007e0 "java",               
    lname=0x4007d8 "openjdk", javaargs=0 '\000', cpwildcard=1 '\001', javaw=0 '\000', ergo=0) at /home/jenkins/workspace/openjdk8_openj9_build_x86-64_linux_largeHeap/openjdk/jdk/src/share/bin/java.c:304                                   
        mode = <optimized out>                                                                                                                                                                                                               
        what = <optimized out>
        cpath = <optimized out>
        main_class = <optimized out>
        ret = 0                                                                                                                                                                                                                              
        ifn = {CreateJavaVM = 0x7f80204708e0 <JNI_CreateJavaVM>, GetDefaultJavaVMInitArgs = 0x7f8020471880 <JNI_GetDefaultJavaVMInitArgs>, GetCreatedJavaVMs = 0x7f8020471860 <JNI_GetCreatedJavaVMs>}                                       
        start = <optimized out>                                                                                                                                                                                                              
        end = <optimized out>
        jvmpath = "/usr/lib/jvm/openj9/jre/lib/amd64/j9vm/libjvm.so", '\000' <repeats 4047 times>                                                                                                                                            
        jrepath = "/usr/lib/jvm/openj9/jre\000java", '\000' <repeats 4067 times>                                                                                                                                                             
        jvmcfg = "/usr/lib/jvm/openj9/jre/lib/amd64/jvm.cfg", '\000' <repeats 3279 times>...                                                                                                                                                 
#6  0x000000000040060c in main ()                                                                                                                                                                                                            
No symbol table info available.
```

**Comment by akrus** [comment]
Ubuntu 18.04.1, kernel version is 4.15.0-38. It's just a few minor releases behind.

**Comment by dmitripivkine** [comment]
Would you please provide full native stacks for all threads `thread apply all bt` ?

**Comment by akrus** [comment]
@dmitripivkine, sorry for late response, file is attached.
[gdb.txt.gz](https://github.com/eclipse/openj9/files/2731977/gdb.txt.gz)

**Comment by dmitripivkine** [comment]
@akrus Do you have matching javacore? I am looking for JVM public flags for each thread. If you don't have matching javacore or you can not share it with us you can get this information using `<Java_Home>/bin/jdmpview -core <core_dump>` as output from `!threads flags` command.

**Comment by dmitripivkine** [comment]
Seems GC Master thread is waiting in `managementGlobalGCStart ()`

```
Thread 19 (Thread 0x7f53e79b7700 (LWP 31570)):
#0  0x00007f8020e949f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f54f9608d04) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f53f0004f58, cond=0x7f54f9608cd8) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x7f54f9608cd8, mutex=0x7f53f0004f58) at pthread_cond_wait.c:655
#3  0x00007f801eec7d3b in omrthread_monitor_wait () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#4  0x00007f801eed1918 in omrthread_rwmutex_enter_write () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#5  0x00007f8017b631db in gcStartEvent () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libjclse7b_29.so
#6  0x00007f8017b63621 in managementGlobalGCStart () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libjclse7b_29.so
#7  0x00007f801ecbb387 in J9HookDispatch () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9hookable29.so
#8  0x00007f801c69b951 in MM_RealtimeGC::reportGCStart(MM_EnvironmentBase*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#9  0x00007f801c6a3c6f in MM_Scheduler::reportStartGCIncrement(MM_EnvironmentRealtime*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#10 0x00007f801c6a4519 in MM_Scheduler::yieldFromGC(MM_EnvironmentRealtime*, bool) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#11 0x00007f801c6a46bb in MM_Scheduler::condYieldFromGC(MM_EnvironmentBase*, unsigned long) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#12 0x00007f801c772feb in MM_SweepSchemeSegregated::incrementalSweepSmall(MM_EnvironmentBase*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#13 0x00007f801c77313f in MM_SweepSchemeSegregated::sweep(MM_EnvironmentBase*, MM_MemoryPoolSegregated*, bool) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#14 0x00007f801c74a2a7 in MM_Dispatcher::run(MM_EnvironmentBase*, MM_Task*, unsigned long) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#15 0x00007f801c69b6aa in MM_RealtimeGC::incrementalCollect(MM_EnvironmentRealtime*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#16 0x00007f801c69b829 in MM_RealtimeGC::internalGarbageCollect(MM_EnvironmentBase*, MM_MemorySubSpace*, MM_AllocateDescription*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#17 0x00007f801c748917 in MM_Collector::garbageCollect(MM_EnvironmentBase*, MM_MemorySubSpace*, MM_AllocateDescription*, unsigned int, MM_ObjectAllocationInterface*, MM_MemorySubSpace*, MM_AllocationContext*) ()
   from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#18 0x00007f801c69361f in MM_MemorySubSpaceMetronome::collect(MM_EnvironmentBase*, MM_GCCode) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#19 0x00007f801c6a437e in MM_Scheduler::masterEntryPoint(MM_EnvironmentBase*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#20 0x00007f801c761050 in dispatcher_thread_proc2(OMRPortLibrary*, void*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#21 0x00007f801ea7880d in omrsig_protect () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9prt29.so
#22 0x00007f801c760e4f in dispatcher_thread_proc () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#23 0x00007f801eecd007 in thread_wrapper () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#24 0x00007f8020e8e6db in start_thread (arg=0x7f53e79b7700) at pthread_create.c:463
#25 0x00007f802079d88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
for
```
Thread 14 (Thread 0x7f5514074700 (LWP 31550)):
#0  0x00007f8020e949f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f54f9317d2c) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f53f0005008, cond=0x7f54f9317d00) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x7f54f9317d00, mutex=0x7f53f0005008) at pthread_cond_wait.c:655
#3  0x00007f801eec7d3b in omrthread_monitor_wait () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#4  0x00007f8017b663ec in Java_com_ibm_lang_management_internal_MemoryNotificationThread_processNotificationLoop () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libjclse7b_29.so
#5  0x00007f801f211054 in ffi_call_unix64 () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#6  0x00007f801f20feed in ffi_call () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#7  0x00007f801f105154 in VM_BytecodeInterpreter::run(J9VMThread*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#8  0x00007f801f0f1dbf in bytecodeLoop () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#9  0x00007f801f1ab222 in c_cInterpreter () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#10 0x00007f801f15b8e4 in runJavaThread () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#11 0x00007f801f1aab76 in javaProtectedThreadProc () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#12 0x00007f801ea7880d in omrsig_protect () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9prt29.so
#13 0x00007f801f1a73a9 in javaThreadProc () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#14 0x00007f801eecd007 in thread_wrapper () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#15 0x00007f8020e8e6db in start_thread (arg=0x7f5514074700) at pthread_create.c:463
#16 0x00007f802079d88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
@LinHu2016 Is it looks like an issue you fixed recently?

**Comment by DanHeidinga** [comment]
@akrus Can you add the `java -version` output from the OpenJ9 build you're using?  

This looks very similar to the issue @LinHu2016 fixed: https://github.com/eclipse/openj9/pull/3060

**Comment by LinHu2016** [comment]
@DanHeidinga @dmitripivkine it might not be directly related with the fix #3060, for fix #3060, the hang thread should call the java api getLastGCInfo().
base on @dmitripivkine 's comment
Thread 14, which is native notification thread, looks normal (wait for the notification queue)
for Thread 19 it is in gc start hook method gcStartEvent(), looks like it aquires and is waiting mgmt->managementDataLock for updating vm management data.

**Comment by dmitripivkine** [comment]
1) Please be aware that your JVM does not have fix for #3060 and maybe it make sense to use later build (however there is no confirmation that you have this particular problem)
2) Thank you for threads flags list. However I apologize but would you please also send me output for `!threads` from `jdmpview`. I realized that native debugger output uses threads `tid` for reference but flags table uses `j9VMThreads` addresses. The output from `!threads` should have both and help me to match.

>@dmitripivkine , core dump is ~200G and as it might contain some sensitive information, not sure if I can pass it directly.

I am sorry for misunderstanding. I never requested `core.dmp` file (which is huge and might contain sensitive information). Instead I asked about `javacore.txt` file (which is text, not too large and likely does not contain sensitive information).

**Comment by akrus** [comment]
1. Yes, I didn't update any systems yet because of #3413 (so waiting for 0.12).
2. Sent threads output.
Unfortunately there is no javacore.txt file as the process didn't crash and I collected the dump using gdump. Can I generate javacore.txt somehow using core dump?

**Comment by dmitripivkine** [comment]
> Unfortunately there is no javacore.txt file as the process didn't crash and I collected the dump using gdump. Can I generate javacore.txt somehow using core dump?

It is ok, most likely jdmpview should show correct values. Unfortunately there is no way to generate javacore from core dump

**Comment by dmitripivkine** [comment]
I can confirm this is exactly the problem fixed in #3060

GC Master Thread:
```
!j9vmthread 0x7f53d0003c00 publicFlags=20 privateFlags=80001a // GC Slave
```
has an Exclusive access (other threads are parked). However according native stack
```
#0  0x00007f8020e949f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f54f9608d04) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f53f0004f58, cond=0x7f54f9608cd8) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x7f54f9608cd8, mutex=0x7f53f0004f58) at pthread_cond_wait.c:655
#3  0x00007f801eec7d3b in omrthread_monitor_wait () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#4  0x00007f801eed1918 in omrthread_rwmutex_enter_write () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#5  0x00007f8017b631db in gcStartEvent () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libjclse7b_29.so
#6  0x00007f8017b63621 in managementGlobalGCStart () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libjclse7b_29.so
#7  0x00007f801ecbb387 in J9HookDispatch () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9hookable29.so
#8  0x00007f801c69b951 in MM_RealtimeGC::reportGCStart(MM_EnvironmentBase*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#9  0x00007f801c6a3c6f in MM_Scheduler::reportStartGCIncrement(MM_EnvironmentRealtime*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#10 0x00007f801c6a4519 in MM_Scheduler::yieldFromGC(MM_EnvironmentRealtime*, bool) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#11 0x00007f801c6a46bb in MM_Scheduler::condYieldFromGC(MM_EnvironmentBase*, unsigned long) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#12 0x00007f801c772feb in MM_SweepSchemeSegregated::incrementalSweepSmall(MM_EnvironmentBase*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#13 0x00007f801c77313f in MM_SweepSchemeSegregated::sweep(MM_EnvironmentBase*, MM_MemoryPoolSegregated*, bool) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#14 0x00007f801c74a2a7 in MM_Dispatcher::run(MM_EnvironmentBase*, MM_Task*, unsigned long) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#15 0x00007f801c69b6aa in MM_RealtimeGC::incrementalCollect(MM_EnvironmentRealtime*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#16 0x00007f801c69b829 in MM_RealtimeGC::internalGarbageCollect(MM_EnvironmentBase*, MM_MemorySubSpace*, MM_AllocateDescription*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#17 0x00007f801c748917 in MM_Collector::garbageCollect(MM_EnvironmentBase*, MM_MemorySubSpace*, MM_AllocateDescription*, unsigned int, MM_ObjectAllocationInterface*, MM_MemorySubSpace*, MM_AllocationContext*) ()
   from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#18 0x00007f801c69361f in MM_MemorySubSpaceMetronome::collect(MM_EnvironmentBase*, MM_GCCode) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#19 0x00007f801c6a437e in MM_Scheduler::masterEntryPoint(MM_EnvironmentBase*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#20 0x00007f801c761050 in dispatcher_thread_proc2(OMRPortLibrary*, void*) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#21 0x00007f801ea7880d in omrsig_protect () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9prt29.so
#22 0x00007f801c760e4f in dispatcher_thread_proc () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#23 0x00007f801eecd007 in thread_wrapper () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#24 0x00007f8020e8e6db in start_thread (arg=0x7f53e79b7700) at pthread_create.c:463
#25 0x00007f802079d88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
it is in Hook code for `managementGlobalGCStart()` and waiting on monitor (forever because owner of the monitor is parked). The reason is in another thread:
```
0  0x00007f8020e949f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f534c124c84) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f52181879e8, cond=0x7f534c124c58) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x7f534c124c58, mutex=0x7f52181879e8) at pthread_cond_wait.c:655
#3  0x00007f801eec7d3b in omrthread_monitor_wait () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9thr29.so
#4  0x00007f801f154c94 in internalAcquireVMAccessNoMutexWithMask () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#5  0x00007f801f1549cc in internalAcquireVMAccess () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#6  0x00007f801c6a3925 in MM_Scheduler::continueGC(MM_EnvironmentRealtime*, GCReason, unsigned long, J9VMThread*, bool) () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#7  0x00007f801c671fbc in J9AllocateObject () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#8  0x00007f801c665927 in j9gc_createJavaLangString () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9gc29.so
#9  0x00007f801f17f32b in newStringUTF () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libj9vm29.so
#10 0x00007f8017b6290b in Java_com_ibm_lang_management_internal_ExtendedGarbageCollectorMXBeanImpl_getLastGcInfoImpl () from /usr/lib/jvm/openj9/jre/lib/amd64/default/libjclse7b_29.so
```
which actually is an initiator of GC (through attempt to allocate object). 

This is exactly problem has been fixed by @LinHu2016.
The fix has been submitted August 22nd, so hopefully with newer version this problem should never happen again.

**Comment by akrus** [comment]
@dmitripivkine , good. I'll update to 0.11 this week then.

**Comment by akrus** [comment]
Looks OK with 0.11, closing.

**Comment by DanHeidinga** [comment]
Thanks for confirming @akrus and for all you help with reproducing this issue!  I'm glad we were able to get it resolved.
